### PR TITLE
Postgres and mysql monitors fix

### PR DESCRIPTION
**Pull request by ArthurKamalov**
* Fix postgres_monitor error caused by wrong type conversion during the parsing of the config file.
* Fall back to native json library if orjson can not handle unsigned 64bit integers. 

Monitor tests and log analyzing tools were also improved.

### CODE REVIEW

File: tests/smoke_tests/monitors_test/postgres_test.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+@pytest.fixture()
+def postgres_cursor(postgresql_client):
+    cursor = postgresql_client.cursor()
+
+    yield cursor
+
+    cursor.close()
+
+
+class PostgresAgentRunner(AgentRunner):
```
Comment (Kami) [inline tests/smoke_tests/monitors_test/postgres_test.py:71]: Great, thanks for working on those tests :+1:

File: tests/smoke_tests/monitors_test/postgres_test.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+USERNAME = "scalyr_test_user"
+PASSWORD = "scalyr_test_password"
+DATABASE = "scalyr_test_db"
+
+
+@pytest.fixture()
+def postgresql_client():
+    os.system("service postgresql start")
+
+    os.system(
```
Comment (Kami) [inline tests/smoke_tests/monitors_test/postgres_test.py:48]: I know this code is only used in the test context and we use hard-coded values, but still, to be on the safe side and to avoid various static analysis tools to complaining, we should quote all the arguments which get passed to those commands - https://docs.python.org/3/library/shlex.html#shlex.quote (aka username, password).

File: tests/unit/util/json_util_test.py [hunk 3]
```diff
@@ -79,6 +79,9 @@ def test_long(self):
     def test_negative_long(self):
         self.__test_encode_decode(r"-1234567890123456789", -1234567890123456789)
 
+    def test_long_long_int(self):
+        self.__test_encode_decode("18446744073709551615", 18446744073709551615)
```
Comment (Kami) [inline tests/unit/util/json_util_test.py:83]: This check should be fine and cover most of the scenarios, but perhaps to be on the safe we could also add explicit test case for orjson - ensure that it throws overflow error on large int and verify our wrapper works correctly (although, yeah, the second part should already be covered by this test case).

File: tests/utils/log_reader.py [hunk 4]
```diff
@@ -27,77 +28,109 @@
...
+    def wait_for_next_line(self, timeout=10):
+        timeout_time = time.time() + timeout
+        while True:
+            line = self._read_line()
+            if line:
+                return line
+
+            if time.time() >= timeout_time:
+                raise LogReaderTimeoutError(
+                    "Timeout of %s seconds reached while waiting for metrics" % timeout
```
Comment (Kami) [inline tests/utils/log_reader.py:81]: :+1:

File: tests/utils/log_reader.py [hunk 5]
```diff
@@ -27,77 +28,109 @@
...
+                )
+
+            time.sleep(0.01)
+
+    def go_to_end(self):
+        while True:
+            line = self._read_line()
```
Comment (Kami) [inline tests/utils/log_reader.py:88]: IIRC, in some scenarios, underlying ``readline()`` will block until there is data available / EOF is reached, right?

File: tests/utils/log_reader.py [hunk 6]
```diff
@@ -27,77 +28,109 @@
...
+            if not line:
+                break
+
+        pass
```
Comment (Kami) [inline tests/utils/log_reader.py:92]: This pass is redundant/

File: tests/utils/log_reader.py [hunk 7]
```diff
@@ -27,77 +28,109 @@
...
+
+    def wait(self, seconds):
+        """
+        Keep checking for new lines for some period of time in 'seconds'.
+        """
+        stop_time = time.time() + seconds
+        while True:
+            try:
+                self.wait_for_next_line()
+            except LogReaderTimeoutError:
```
Comment (Kami) [inline tests/utils/log_reader.py:118]: Should we log a message here, perhaps under debug?

File: tests/utils/log_reader.py [hunk 8]
```diff
@@ -27,77 +28,109 @@
...
+                break
 
-        with self._wait_for_line_cv:
-            while not self._has_new_line:
-                if time.time() >= timeout_time:
-                    raise ValueError(
-                        "timeout of %s seconds reached while waiting for metrics"
-                        % (timeout)
-                    )
+    def add_error_check(self, pattern, message=None):
```
Comment (Kami) [inline tests/utils/log_reader.py:124]: Please add doc strings (and type annotations) to all the new methods you added or changed.

File: tests/utils/log_reader.py [hunk 9]
```diff
@@ -130,56 +157,67 @@ def _parse_line(self, line):
...
+                return needed_metrics
 
-            if time.time() >= timeout_time:
-                raise ValueError(
-                    "timeout of %s seconds reached while waiting for metrics"
-                    % (timeout)
-                )
+            try:
+                self.wait_for_next_line()
+            except LogReaderTimeoutError():
```
Comment (Kami) [inline tests/utils/log_reader.py:179]: Perhaps also log a message under debug or similar here - something along the line of "Timeout of x seconds reached when waiting for a new line, will retry"

File: tests/utils/log_reader.py [hunk 10]
```diff
@@ -130,56 +157,67 @@ def _parse_line(self, line):
...
+    ):
+        timeout_time = time.time() + timeout
+        while True:
+            try:
+                self.wait_for_next_line()
+            except LogReaderTimeoutError:
+                if time.time() >= timeout_time:
+                    raise
+                else:
+                    continue
```
Comment (Kami) [inline tests/utils/log_reader.py:199]: Same here - let's log a message to make troubleshooting easier.

File: tests/utils/log_reader.py [hunk 11]
```diff
@@ -130,56 +157,67 @@ def _parse_line(self, line):
...
-        return metrics[name]
+    def wait_for_metrics_change(
+        self, metric_names, previous_values, predicate, timeout=10
+    ):
+        timeout_time = time.time() + timeout
+        while True:
+            try:
+                self.wait_for_next_line()
+            except LogReaderTimeoutError:
+                if time.time() >= timeout_time:
```
Comment (Kami) [inline tests/utils/log_reader.py:196]: Perhaps we should throw a different exception here (same in other places where we use the same pattern) so we can differentiate between timeout being hit on wait_for_next_line and global timeout for that method being reached.

**Comment by Kami** [comment]
Once this PR is merged, can you please verify that code coverage data for postgresql monitor is correctly updated?

**Comment by Kami** [comment]
It does appear it's correctly updated and reflected in sonar cloud step - https://app.circleci.com/pipelines/github/scalyr/scalyr-agent-2/4135/workflows/16fac9fa-e4a0-4219-a63c-1e3b5f621ca9/jobs/54561, but for some reason codecov.io data doesn't appear to be getting updated.

I pushed a couple of commits which will hopefully shed some light on why it's not being updated.

**Comment by Kami** [comment]
OK, I fixed it - issue was related to the testing change which was added to get that step to pass in the private repo :)

File: tests/unit/util/json_util_test.py (hunk 3 above)
Comment (ArthurKamalov) [reply tests/unit/util/json_util_test.py:90]: Added orjson.dumps error catch assertions.

File: tests/utils/log_reader.py (hunk 5 above)
Comment (ArthurKamalov) [reply tests/utils/log_reader.py]: readline() does not block if I am not mistaken. It just returns an empty string.

File: tests/utils/log_reader.py (hunk 7 above)
Comment (ArthurKamalov) [reply tests/utils/log_reader.py]: This code part was refactored :)

File: tests/utils/log_reader.py (hunk 8 above)
Comment (ArthurKamalov) [reply tests/utils/log_reader.py]: +

File: tests/utils/log_reader.py (hunk 9 above)
Comment (ArthurKamalov) [reply tests/utils/log_reader.py]: I've added line reading based on generators, so I could remove all those timeout checks.

File: tests/smoke_tests/monitors_test/postgres_test.py (hunk 2 above)
Comment (ArthurKamalov) [reply tests/smoke_tests/monitors_test/postgres_test.py:48]: Thanks for the good advice. Done.

**Comment by ArthurKamalov** [comment]
@Kami part of the code was refactored and something became outdated.

File: tests/smoke_tests/monitors_test/mysql_test.py [hunk 12]
```diff
@@ -153,23 +166,17 @@ def _test(request, python_version):
...
     assert row1_text == "updated_row1"
 
-    reader.wait_for_metrics_increase(metrics_to_check, previous_metric_values)
-
-    current_metrics = reader.get_metrics(metrics_to_check)
-
-    diff = dict(
-        (name, current_metrics[name] - previous_metric_values[name])
-        for name in metrics_to_check
+    reader.wait_for_metrics_equal(
```
Comment (Kami) [inline tests/smoke_tests/monitors_test/mysql_test.py:169]: Cool, I like this approach more :+1:

File: tests/smoke_tests/monitors_test/syslog_test.py [hunk 13]
```diff
@@ -66,28 +66,27 @@ def _test(python_version):
...
 
     os.system("logger message3 --port {0} --udp --server 127.0.0.1".format(UDP_PORT))
 
-    last_line = reader.wait_for_new_line()
+    last_line = reader.wait_for_next_line()
 
     assert "message3" in last_line
 
-    runner.stop()
+    agent_log_reader.go_to_end()
```
Comment (Kami) [inline tests/smoke_tests/monitors_test/syslog_test.py:89]: This works for now, but since we always need to call this ``go_to_end()`` method it may make sense in the future to implement ContextManager which always calls that method on ``__exit__``.

File: tests/utils/log_reader.py [hunk 14]
```diff
@@ -27,101 +32,172 @@
...
+
+    def _line_generator(self):  # type: () -> Generator
+        """Generator that reads all available lines.
+        """
+        for line in self._file:
+            line = line.strip("\n")
+            self._new_line_callback(line)
+            yield line
+
+    def _line_generator_blocking(self, timeout=10):  # type: (float) -> Generator
```
Comment (Kami) [inline tests/utils/log_reader.py:100]: Some tests for this ``LogReader`` class functionality itself would be nice as well :)

**Review by Kami — APPROVED** [review]
LGTM.

When you get a chance, some tests for LogReader functionality would be nice as well.

File: tests/utils/log_reader.py (hunk 14 above)
Comment (ArthurKamalov) [reply tests/utils/log_reader.py:100]: There is already a file `tests/utils/tests/test_log_reader.py ` with some basic tests in it:)
Those tests are invoked by the tox target before the main tests.

File: tests/smoke_tests/monitors_test/syslog_test.py (hunk 13 above)
Comment (ArthurKamalov) [reply tests/smoke_tests/monitors_test/syslog_test.py:89]: Good advice. Added reminder for that.

**Comment by Kami** [comment]
@ArthurKamalov I noticed we are missing corresponding changelog entries - can you please add them.
